File: LR1/Queue/Queue.cpp
#include "Queue.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

// Инициализация очереди с заданной вместимостью
void initQueue(Queue& q, int capacity) {
    //if ()
    q.data = new string[capacity];
    q.size = 0;
    q.capacity = capacity;
    q.front = 0;
    q.back = 0;
}

// Проверка, пуста ли очередь
bool isEmpty(const Queue& q) {
    return q.size == 0;
}

// Добавление элемента в очередь
void Qpush(Queue& q, const string& value) {
    if (q.size >= q.capacity) {
        cout << "Ошибка: очередь переполнена!" << endl;
        return;
    }
    q.data[q.back] = value;
    q.back = (q.back + 1) % q.capacity;
    q.size++;
}

// Удаление элемента из очереди
string Qpop(Queue& q, string& value) {
    if (isEmpty(q)) {
        cout << "Ошибка: очередь пуста!" << endl;
    }
    value = q.data[q.front];
    q.front = (q.front + 1) % q.capacity;
    q.size--;
    return value;
}

// Просмотр первого элемента
string peek(const Queue& q) {
    if (isEmpty(q)) {
        return "Очередь пуста!";
    }
    return q.data[q.front];
}

// Получение длины очереди
int length(const Queue& q) {
    return q.size;
}

// Печать очереди
void print(const Queue& q) {
    if (isEmpty(q)) {
        cout << "Очередь пуста!" << endl;
        return;
    }
    cout << "Очередь: ";
    for (int i = 0, idx = q.front; i < q.size; i++) {
        cout << q.data[idx] << " ";
        idx = (idx + 1) % q.capacity;
    }
    cout << endl;
}

// Очистка очереди
void Qclear(Queue& q) {
    q.size = 0;
    q.front = 0;
    q.back = 0;
}

// Сохранение очереди в файл
void saveQueue(const Queue& q, const string& filename) {
    ofstream file(filename);
    if (!file) {
        cout << "Ошибка открытия файла для записи!" << endl;
        return;
    }
    file << q.size << endl;
    for (int i = 0, idx = q.front; i < q.size; i++) {
        file << q.data[idx] << endl;
        idx = (idx + 1) % q.capacity;
    }
    file.close();
}

// Загрузка очереди из файла
void loadQueue(Queue& q, const string& filename) {
    ifstream file(filename);
    if (!file) {
        cout << "Ошибка открытия файла для чтения!" << endl;
        return;
    }
    Qclear(q);
    int n;
    file >> n;
    file.ignore(); // пропуск символа конца строки
    string value;
    for (int i = 0; i < n; i++) {
        getline(file, value);
        Qpush(q, value);
    }
    file.close();
}

// Очистка памяти очереди
void freeQueue(Queue& q) {
    delete[] q.data;
    q.data = nullptr;
    q.size = 0;
    q.capacity = 0;
    q.front = 0;
    q.back = 0;
}

File: LR1/NaGO/Queue/queue.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

// ===== Структура очереди =====
type Queue struct {
	data     []string
	size     int
	capacity int
	front    int
	back     int
}

// ===== Инициализация очереди =====
func initQueue(q *Queue, capacity int) {
	q.data = make([]string, capacity)
	q.size = 0
	q.capacity = capacity
	q.front = 0
	q.back = 0
}

// ===== Проверка, пуста ли очередь =====
func isEmpty(q *Queue) bool {
	return q.size == 0
}

// ===== Добавление элемента =====
func Qpush(q *Queue, value string) {
	if q.size >= q.capacity {
		fmt.Println("Ошибка: очередь переполнена!")
		return
	}
	q.data[q.back] = value
	q.back = (q.back + 1) % q.capacity
	q.size++
}

// ===== Удаление элемента =====
func Qpop(q *Queue) string {
	if isEmpty(q) {
		fmt.Println("Ошибка: очередь пуста!")
		return ""
	}
	value := q.data[q.front]
	q.front = (q.front + 1) % q.capacity
	q.size--
	return value
}

// ===== Просмотр первого элемента =====
func peek(q *Queue) string {
	if isEmpty(q) {
		return "Очередь пуста!"
	}
	return q.data[q.front]
}

// ===== Получение длины =====
func length(q *Queue) int {
	return q.size
}

// ===== Печать =====
func printQueue(q *Queue) {
	if isEmpty(q) {
		fmt.Println("Очередь пуста!")
		return
	}
	fmt.Print("Очередь: ")
	idx := q.front
	for i := 0; i < q.size; i++ {
		fmt.Print(q.data[idx], " ")
		idx = (idx + 1) % q.capacity
	}
	fmt.Println()
}

// ===== Очистка =====
func Qclear(q *Queue) {
	q.size = 0
	q.front = 0
	q.back = 0
}

// ===== Сохранение в файл =====
func saveQueue(q *Queue, filename string) {
	file, err := os.Create(filename)
	if err != nil {
		fmt.Println("Ошибка открытия файла для записи!")
		return
	}
	defer file.Close()

	writer := bufio.NewWriter(file)
	fmt.Fprintln(writer, q.size)
	idx := q.front
	for i := 0; i < q.size; i++ {
		fmt.Fprintln(writer, q.data[idx])
		idx = (idx + 1) % q.capacity
	}
	writer.Flush()
}

// ===== Загрузка из файла =====
func loadQueue(q *Queue, filename string) {
	file, err := os.Open(filename)
	if err != nil {
		// если файла нет — просто создаём пустую очередь
		return
	}
	defer file.Close()

	Qclear(q)
	scanner := bufio.NewScanner(file)

	if !scanner.Scan() {
		return
	}
	n, _ := strconv.Atoi(scanner.Text())

	for i := 0; i < n && scanner.Scan(); i++ {
		Qpush(q, scanner.Text())
	}
}
